tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,199 @@
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code
+def convert_arguments(args: List[Argument], *, has_tensor_options: bool) -> Tuple[List[Binding], List[str]]:
+    code_list = [f"c10::IValue {args[i].name} = std::move(peek(stack, {i}, {len(args)}));" for i in
+                 range(len(args))] + [""]
+    binding_list = []
+    for i, arg in enumerate(args):
+        unboxed_name, _, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        code_list.extend(code)
+        # we need the 'self' argument so method needs to be False, also it doesn't matter if signature faithful or not
+        arg_binding = argument(arg, cpp_no_default_args=set(), method=False, faithful=False,
+                               has_tensor_options=has_tensor_options)
+        # since no TensorOptions here, we can safely assume there's only 1 element in arg_binding.
+        if not arg_binding:
+            raise Exception(f"Argument {arg} cannot be converted to Binding. has_tensor_options = {has_tensor_options}")
+        binding_list.append(arg_binding[0].with_name(unboxed_name))
+    return binding_list, code_list
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[str, CType, List[str]]:
```

Comment:
Do we actually need to return the `CType` in this function? It looks like it's only used in one place and ignored everywhere else.

Why not just compute it directly in the one place where it's needed? (in `_gen_code_list_type`)

Reply:
On a similar note, can you add in the comment above this file what everything that it returns corresponds to?

It would be kind of nice if we could just take in an `Argument` instead of the separate componeents of it (type / name / mutability), but I guess you need that in order to make it recursively work on nested types.

"Takes in the type, name and mutability corresponding to an argument, and generates a tuple of:
(1) the C++ code necessary to unbox the argument
(2) A `Binding` corresponding to the newly created unboxed variable
"

Reply:
The `CType` is still being returned here? Why do we need it? (see above comment - can't we just generate outside of the function in one place where its needed)

Reply:
We need `CType` for list element. If we generate outside of the function it will only be a type of List but not the type of the element.

Reply:
Ok. Mechanically it looks like you can't because you don't have e.g the `is_mutable` info by the time you're in the list function.

I think this is fine good enough shape for now. I could imagine another way of doing this by passing the `Argument` object around everywhere, but you might run into some issues doing that too. No more action needed :)